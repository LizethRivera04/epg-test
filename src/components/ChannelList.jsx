import Proptypes from "prop-types";

export const ChannelList = ({ channel }) => {
  const { number, name, image } = channel;
  return (
    <>
      <div className="">
        <div className="flex max-w-32 drop-shadow-lg justify-between">
          <p>{number}</p>
          <img className="max-w-20" src={image} alt={name} />
        </div>
      </div>
    </>
  );
};

ChannelList.propTypes = {
  channel: Proptypes.object.isRequired,
};
