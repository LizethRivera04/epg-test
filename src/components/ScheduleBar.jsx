import Proptypes from "prop-types";

import dayjs from "dayjs";
import uniqid from "uniqid";
import { getDaysBetween } from "../utils/scheduleData";

export const ScheduleBar = ({ channels }) => {
  const start = dayjs(channels[0].events[0].date_begin);
  // start.setHours(9, 0, 0, 0);
  const end = dayjs(
    channels[channels.length - 1].events[
      channels[channels.length - 1].events.length - 1
    ].date_end
  );
  const scheduleData = getDaysBetween(start, end);

  return (
    <div className="flex h-11 w-fit bg-gray-800">
      <div className=" py-2 w-[calc(100%-160px)] flex ">
        {scheduleData.map((schedule) => (
          <div className="px-11" key={uniqid()}>
            {schedule}
          </div>
        ))}
      </div>
    </div>
  );
};

ScheduleBar.propTypes = {
  channels: Proptypes.array.isRequired,
};
