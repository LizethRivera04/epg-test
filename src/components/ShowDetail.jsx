import Proptypes from "prop-types";

import dayjs from "dayjs";

export const ShowDetail = ({ showDetail }) => {
  const { event } = showDetail;
  const dateBegin = dayjs(event.date_begin);
  const dateEnd = dayjs(event.date_end);
  return (
    <div className="flex justify-start flex-col xs:w-full md:w-1/2 p-4 md:p-5 items-start">
      <h1>{event.name}</h1>
      <div className="mt-5 text-start">
        <p aria-label="schedule-tag">
          {dayjs(dateBegin).format("HH:mm")}hrs a{" "}
          {dayjs(dateEnd).format("HH:mm")}hrs {dateBegin.diff(dateEnd, "hours")}
          hr
        </p>
        <p className="">
          Veniam reprehenderit officia ex exercitation dolore aute pariatur
          labore magna eu pariatur. Quis laborum amet nostrud sit quis. Occaecat
          aute voluptate aliquip labore ipsum qui occaecat et cillum id.
          Excepteur ea ea fugiat eiusmod quis dolore elit mollit elit. Non sunt
          sit ex ipsum consectetur esse et dolor sint anim ipsum Lorem
          reprehenderit. Cillum nulla ea commodo qui sunt minim exercitation
          consequat fugiat sunt ad duis ipsum. Aliquip elit qui ad officia magna
          commodo eu ex ipsum fugiat.
        </p>
      </div>
    </div>
  );
};

ShowDetail.propTypes = {
  showDetail: Proptypes.object.isRequired,
};
