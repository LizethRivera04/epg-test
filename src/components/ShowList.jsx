import Proptypes from "prop-types";
import dayjs from "dayjs";
import uniqid from "uniqid";

export const ShowList = ({ events, setShowDetail }) => {
  return (
    <>
      <div className="flex w-fit">
        {events.map((event) => (
          <div
            aria-label="div-show"
            key={uniqid()}
            className="p-8 border-gray-300 border-solid border-[1px] w-64 cursor-pointer"
            onMouseEnter={() => setShowDetail({ show: true, event: event })}
          >
            <p>{event.name}</p>
            <span>{dayjs(event.date_begin).format("HH:mm")}</span>-
            <span>{dayjs(event.date_end).format("HH:mm")}</span>
          </div>
        ))}
      </div>
    </>
  );
};

ShowList.propTypes = {
  events: Proptypes.array.isRequired,
  setShowDetail: Proptypes.func.isRequired,
};
