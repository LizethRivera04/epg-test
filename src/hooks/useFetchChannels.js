import { useEffect, useMemo, useState } from "react";
import { dataApi } from "../api/data";

const useFetchChannels = () => {
  const [channels, setChannels] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  const fetchDataList = useMemo(
    () => async () => {
      try {
        const data = await dataApi();
        setChannels(data);
        setIsLoading(false);
      } catch (error) {
        console.error("Error fetching data list:", error);
      }
    },
    []
  );

  useEffect(() => {
    if (channels.length === 0) {
      fetchDataList();
    }
  }, [fetchDataList]);

  return {
    channels,
    isLoading,
  };
};

export default useFetchChannels;
