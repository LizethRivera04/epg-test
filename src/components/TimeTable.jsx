import { useState } from "react";

import Proptypes from "prop-types";
import uniqid from "uniqid";

import useFetchChannels from "../hooks/useFetchChannels";
import { ShowDetail } from "./ShowDetail";
import { ShowList } from "./ShowList";
import { ChannelList } from "./ChannelList";
import { ScheduleBar } from "./ScheduleBar";
import { IconSpinner } from "./layout/IconSpinner";

import { IconTimes } from "./layout/IconTimes";

export const TimeTable = ({ setShowModal }) => {
  const [showDetail, setShowDetail] = useState({
    show: false,
    event: {},
  });

  const { channels, isLoading } = useFetchChannels();
  return (
    <div
      id="default-modal"
      tabIndex="-1"
      aria-hidden="true"
      className=" overflow-y-auto overflow-x-hidden fixed top-0 right-0 left-0 z-50 justify-center items-center w-full md:inset-0 h-[calc(100%-1rem)] max-h-full"
    >
      <div className="relative p-4 w-full h-[calc(100%-1rem)] ">
        <div className="relative bg-white h-full rounded-lg shadow dark:bg-gray-700 overflow-x-auto">
          <div
            aria-label="div-close-modal"
            className="w-full flex justify-end"
            onClick={() => setShowModal(false)}
          >
            <button
              type="button"
              className="text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm w-12 h-12 ms-auto inline-flex justify-center items-center dark:hover:bg-gray-600 dark:hover:text-white"
              data-modal-hide="default-modal"
            >
              <IconTimes />

              <span className="sr-only">Close modal</span>
            </button>
          </div>
          {isLoading && <IconSpinner />}

          {showDetail.show && (
            <div className="flex items-center justify-between p-4 md:p-5 border-b rounded-t dark:border-gray-600">
              <ShowDetail showDetail={showDetail} />
            </div>
          )}

          <div className="p-4 pt-0 md:p-0 md:pt-0 flex">
            <div className="pt-[44px]">
              {!isLoading &&
                channels.map((channel) => (
                  <div key={uniqid()} className="">
                    <div className="p-4 md:p-5 bg-gray-600 border-gray-900 border-solid flex items-center border-r-2 xs:w-24 xs:h-28 md:w-40">
                      <ChannelList channel={channel} />
                    </div>
                  </div>
                ))}
            </div>

            <div id="scroll-channels" className="overflow-y-auto">
              {!isLoading && (
                <>
                  <div id="schedule">
                    <ScheduleBar channels={channels} />
                  </div>

                  {channels.map((channel) => (
                    <div key={uniqid()} className="">
                      <div className="">
                        <ShowList
                          key={uniqid()}
                          events={channel.events}
                          setShowDetail={setShowDetail}
                        />
                      </div>
                    </div>
                  ))}
                </>
              )}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

TimeTable.propTypes = {
  setShowModal: Proptypes.func.isRequired,
};
