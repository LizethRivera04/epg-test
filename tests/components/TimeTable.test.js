import { describe, expect, test } from "@jest/globals";
import { TimeTable } from "../../src/components/TimeTable";
import { fireEvent, render, screen } from "@testing-library/react";
import useFetchChannels from "../../src/hooks/useFetchChannels";

jest.mock("../../src/hooks/useFetchChannels");
describe("Test in <TimeTable/>", () => {
  const setShowModal = jest.fn();

  test("should to show loading", () => {
    useFetchChannels.mockReturnValue({
      channels: [],
      isLoading: true,
    });
    render(<TimeTable setShowModal={setShowModal} />);
    expect(screen.getByText("Cargando...")).toBeTruthy();
  });

  test("should to match with snapshot", () => {
    const { container } = render(<TimeTable setShowModal={setShowModal} />);
    expect(container).toMatchSnapshot();
  });

  test("IconTimes should to call setShowModal function", () => {
    render(<TimeTable setShowModal={setShowModal} />);

    const divCloseModal = screen.getByLabelText("div-close-modal");
    fireEvent.click(divCloseModal);
    expect(setShowModal).toHaveBeenCalled();
    expect(setShowModal).toHaveBeenCalledWith(false);
  });
});
