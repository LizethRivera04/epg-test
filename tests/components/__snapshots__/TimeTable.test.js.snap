// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <TimeTable/> should to match with snapshot 1`] = `
<div>
  <div
    aria-hidden="true"
    class=" overflow-y-auto overflow-x-hidden fixed top-0 right-0 left-0 z-50 justify-center items-center w-full md:inset-0 h-[calc(100%-1rem)] max-h-full"
    id="default-modal"
    tabindex="-1"
  >
    <div
      class="relative p-4 w-full h-[calc(100%-1rem)] "
    >
      <div
        class="relative bg-white h-full rounded-lg shadow dark:bg-gray-700 overflow-x-auto"
      >
        <div
          aria-label="div-close-modal"
          class="w-full flex justify-end"
        >
          <button
            class="text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm w-12 h-12 ms-auto inline-flex justify-center items-center dark:hover:bg-gray-600 dark:hover:text-white"
            data-modal-hide="default-modal"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="w-4 h-4 text-gray-100 dark:text-white"
              fill="none"
              viewBox="0 0 14 14"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <span
              class="sr-only"
            >
              Close modal
            </span>
          </button>
        </div>
        <div
          class="w-full h-[calc(100%-4rem)] flex flex-col justify-center items-center"
        >
          <svg
            class="animate-spin h-12 w-12 text-gray-300"
            fill="none"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              class="opacity-25"
              cx="12"
              cy="12"
              r="10"
              stroke="currentColor"
              stroke-width="4"
            />
            <path
              class="opacity-75"
              d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
              fill="currentColor"
            />
          </svg>
          <span
            class="mt-2"
          >
            Cargando...
          </span>
        </div>
        <div
          class="p-4 pt-0 md:p-0 md:pt-0 flex"
        >
          <div
            class="pt-[44px]"
          />
          <div
            class="overflow-y-auto"
            id="scroll-channels"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
